File: integrations/openclaw/src/tools.ts
import { randomUUID } from "node:crypto";
import { realpath, stat } from "node:fs/promises";
import { isAbsolute, relative, resolve, sep } from "node:path";
import type { Agent, WorkroomDetail } from "@atalk/sdk";
import type { OpenClawPluginApi } from "openclaw/plugin-sdk/channel-core";
import { Type } from "typebox";
import { getActiveAtalkAgent } from "./channel.js";

const MAX_ATTACHMENT_BYTES = 100 * 1024 * 1024;
const uuid = () => Type.String({ format: "uuid" });
const handle = () => Type.String({ pattern: "^@[a-z0-9][a-z0-9._-]{1,62}$" });
const handles = () => Type.Array(handle(), { maxItems: 100, default: [] });
const common = {
  workroomId: uuid(),
  threadId: uuid(),
  operationId: Type.Optional(uuid()),
  mandateId: Type.Optional(uuid()),
  rationale: Type.Optional(Type.String({ minLength: 1, maxLength: 4_000 })),
};

function jsonResult(value: unknown) {
  return {
    content: [{ type: "text" as const, text: JSON.stringify(value, null, 2) }],
    details: value,
  };
}

function operationId(value?: string): string {
  return value ?? randomUUID();
}

function taskView(detail: WorkroomDetail) {
  return {
    id: detail.workroom.id,
    status: detail.workroom.status,
    deadline: detail.workroom.deadline ?? null,
    descriptor: detail.descriptor,
    membership: {
      role: detail.membership.role,
      joinedAt: detail.membership.joinedAt,
      leftAt: detail.membership.leftAt ?? null,
    },
    members: detail.members.map(({ membership, peer }) => ({
      id: membership.peerId,
      role: membership.role,
      leftAt: membership.leftAt ?? null,
      handle: peer?.handle ?? null,
      displayName: peer?.displayName ?? null,
      type: peer?.type ?? null,
    })),
    threads: detail.threads.map((thread) => ({ id: thread.id, kind: thread.kind, createdAt: thread.createdAt })),
    permissions: detail.latestMandates.map(({ mandate, revocation }) => ({
      mandateId: mandate.mandateId,
      revision: mandate.revision,
      actorPeerId: mandate.actorPeerId,
      validFrom: mandate.validFrom,
      validUntil: mandate.validUntil,
      revoked: Boolean(revocation),
    })),
    approvals: detail.approvals.map((approval) => ({
      requestId: approval.requestId,
      status: approval.status,
      requiredApprovals: approval.requiredApprovals,
      eligiblePeerIds: approval.eligiblePeerIds,
      expiresAt: approval.expiresAt ?? null,
    })),
  };
}

function participantIds(detail: WorkroomDetail): string[] {
  return detail.members
    .filter(({ membership }) => !membership.leftAt)
    .map(({ membership }) => membership.peerId);
}

function resolveMembers(detail: WorkroomDetail, requestedHandles: string[]) {
  const activeMembers = detail.members.filter(({ membership }) => !membership.leftAt && membership.peerId);
  return requestedHandles.map((requested) => {
    const member = activeMembers.find(({ peer }) => peer?.handle === requested);
    if (!member?.peer) throw new Error(`Active Task participant not found: ${requested}`);
    return member.peer;
  });
}

function mentions(detail: WorkroomDetail, requestedHandles: string[]) {
  return resolveMembers(detail, requestedHandles).map((peer) => ({
    peerId: peer.id,
    handle: peer.handle,
    peerType: peer.type === "AGENT" ? "AGENT" as const : "HUMAN" as const,
    intent: "direct" as const,
  }));
}

async function workspaceFile(filePath: string, workspaceDir: string) {
  const root = await realpath(resolve(workspaceDir));
  const path = await realpath(resolve(workspaceDir, filePath));
  const fromRoot = relative(root, path);
  if (fromRoot === ".." || fromRoot.startsWith(`..${sep}`) || isAbsolute(fromRoot)) {
    throw new Error("Task files must stay inside the active OpenClaw workspace");
  }
  const metadata = await stat(path);
  if (!metadata.isFile()) throw new Error("Task attachment path is not a regular file");
  if (metadata.size > MAX_ATTACHMENT_BYTES) throw new Error("aTalk Task attachments cannot exceed 100 MB");
  return path;
}

export function buildAtalkTaskTools(agent: Agent, context: { workspaceDir?: string }) {
  const tools = [
      {
        name: "atalk_task_list",
        label: "List aTalk Tasks",
        description: "List encrypted aTalk Tasks assigned to this agent. Titles and objectives are verified and decrypted locally.",
        parameters: Type.Object({
          cursor: Type.Optional(Type.String()),
          limit: Type.Optional(Type.Integer({ minimum: 1, maximum: 200, default: 50 })),
        }, { additionalProperties: false }),
        async execute(_toolCallId: string, params: { cursor?: string; limit?: number }) {
          const page = await agent.workrooms.list(params.cursor, params.limit ?? 50);
          return jsonResult({ tasks: page.workrooms.map(taskView), nextCursor: page.nextCursor });
        },
      },
      {
        name: "atalk_task_open",
        label: "Open an aTalk Task",
        description: "Open one encrypted aTalk Task and inspect its verified objective, participants, threads, permission metadata and approval status.",
        parameters: Type.Object({ workroomId: uuid() }, { additionalProperties: false }),
        async execute(_toolCallId: string, params: { workroomId: string }) {
          return jsonResult(taskView(await agent.workrooms.get(params.workroomId, 0, 1)));
        },
      },
      {
        name: "atalk_task_message",
        label: "Message an aTalk Task",
        description: "Publish an encrypted Task message under the agent's current permission. Mention handles to direct it to specific humans or agents.",
        parameters: Type.Object({
          ...common,
          body: Type.String({ minLength: 1, maxLength: 200_000 }),
          mentionHandles: Type.Optional(handles()),
          replyToEventId: Type.Optional(uuid()),
        }, { additionalProperties: false }),
        async execute(_toolCallId: string, params: {
          workroomId: string; threadId: string; operationId?: string; mandateId?: string; rationale?: string;
          body: string; mentionHandles?: string[]; replyToEventId?: string;
        }) {
          const id = operationId(params.operationId);
          const detail = await agent.workrooms.get(params.workroomId, 0, 1);
          const result = await agent.workrooms.publishMandated({
            workroomId: params.workroomId,
            threadId: params.threadId,
            operationId: id,
            ...(params.mandateId ? { mandateId: params.mandateId } : {}),
            ...(params.rationale ? { rationale: params.rationale } : {}),
            participantPeerIds: participantIds(detail),
            payload: {
              version: 1,
              kind: "message",
              threadId: params.threadId,
              body: params.body,
              mentions: mentions(detail, params.mentionHandles ?? []),
              ...(params.replyToEventId ? { replyToEventId: params.replyToEventId } : {}),
            },
          });
          return jsonResult({ operationId: id, result });
        },
      },
      {
        name: "atalk_task_activity",
        label: "Update aTalk Task activity",
        description: "Publish a concise encrypted progress update under the agent's current permission.",
        parameters: Type.Object({
          ...common,
          activityType: Type.String({ minLength: 1, maxLength: 160 }),
          summary: Type.String({ minLength: 1, maxLength: 4_000 }),
          mentionHandles: Type.Optional(handles()),
        }, { additionalProperties: false }),
        async execute(_toolCallId: string, params: {
          workroomId: string; threadId: string; operationId?: string; mandateId?: string; rationale?: string;
          activityType: string; summary: string; mentionHandles?: string[];
        }) {
          const id = operationId(params.operationId);
          const detail = await agent.workrooms.get(params.workroomId, 0, 1);
          const result = await agent.workrooms.publishMandated({
            workroomId: params.workroomId,
            threadId: params.threadId,
            operationId: id,
            ...(params.mandateId ? { mandateId: params.mandateId } : {}),
            ...(params.rationale ? { rationale: params.rationale } : {}),
            participantPeerIds: participantIds(detail),
            payload: {
              version: 1,
              kind: "activity",
              threadId: params.threadId,
              activityType: params.activityType,
              summary: params.summary,
              mentions: mentions(detail, params.mentionHandles ?? []),
              sourceEventIds: [],
              attributes: {},
            },
          });
          return jsonResult({ operationId: id, result });
        },
      },
      {
        name: "atalk_task_plan",
        label: "Plan an aTalk Task",
        description: "Publish a versioned encrypted Task plan. Assign steps by active participant handle. Requires the explicit plan.update permission.",
        parameters: Type.Object({
          ...common,
          planId: Type.Optional(uuid()),
          planVersion: Type.Integer({ minimum: 1 }),
          summary: Type.String({ minLength: 1, maxLength: 2_000 }),
          steps: Type.Array(Type.Object({
            id: Type.String({ minLength: 1, maxLength: 160 }),
            title: Type.String({ minLength: 1, maxLength: 500 }),
            status: Type.Union([
              Type.Literal("executing"), Type.Literal("waiting_approval"), Type.Literal("blocked"),
              Type.Literal("completed"), Type.Literal("cancelled"), Type.Literal("expired"),
            ]),
            assignedHandles: Type.Optional(handles()),
            dependsOnStepIds: Type.Optional(Type.Array(Type.String({ minLength: 1, maxLength: 160 }), { maxItems: 100 })),
            deadline: Type.Optional(Type.String({ format: "date-time" })),
          }, { additionalProperties: false }), { minItems: 1, maxItems: 500 }),
        }, { additionalProperties: false }),
        async execute(_toolCallId: string, params: {
          workroomId: string; threadId: string; operationId?: string; mandateId?: string; rationale?: string;
          planId?: string; planVersion: number; summary: string;
          steps: Array<{ id: string; title: string; status: "executing" | "waiting_approval" | "blocked" | "completed" | "cancelled" | "expired"; assignedHandles?: string[]; dependsOnStepIds?: string[]; deadline?: string }>;
        }) {
          const id = operationId(params.operationId);
          const detail = await agent.workrooms.get(params.workroomId, 0, 1);
          const result = await agent.workrooms.publishMandated({
            workroomId: params.workroomId,
            threadId: params.threadId,
            operationId: id,
            ...(params.mandateId ? { mandateId: params.mandateId } : {}),
            ...(params.rationale ? { rationale: params.rationale } : {}),
            participantPeerIds: participantIds(detail),
            payload: {
              version: 1,
              kind: "plan",
              ...(params.planId ? { planId: params.planId } : {}),
              planVersion: params.planVersion,
              summary: params.summary,
              steps: params.steps.map((step) => ({
                id: step.id,
                title: step.title,
                status: step.status,
                assignedPeerIds: resolveMembers(detail, step.assignedHandles ?? []).map(({ id: peerId }) => peerId),
                dependsOnStepIds: step.dependsOnStepIds ?? [],
                ...(step.deadline ? { deadline: step.deadline } : {}),
              })),
            },
          });
          return jsonResult({ operationId: id, result });
        },
      },
      {
        name: "atalk_task_deliverable",
        label: "Submit an aTalk Task deliverable",
        description: "Submit an encrypted artifact version for review under the agent's current deliverable.submit permission.",
        parameters: Type.Object({
          ...common,
          artifactId: uuid(),
          artifactVersion: Type.Integer({ minimum: 1 }),
          artifactVersionId: uuid(),
          deliverableId: Type.Optional(uuid()),
          acceptanceCriteria: Type.Array(Type.String({ minLength: 1, maxLength: 1_000 }), { minItems: 1, maxItems: 100 }),
          note: Type.Optional(Type.String({ maxLength: 4_000 })),
          mentionHandles: Type.Optional(handles()),
        }, { additionalProperties: false }),
        async execute(_toolCallId: string, params: {
          workroomId: string; threadId: string; operationId?: string; mandateId?: string; rationale?: string;
          artifactId: string; artifactVersion: number; artifactVersionId: string; deliverableId?: string;
          acceptanceCriteria: string[]; note?: string; mentionHandles?: string[];
        }) {
          const id = operationId(params.operationId);
          const detail = await agent.workrooms.get(params.workroomId, 0, 1);
          const result = await agent.workrooms.publishMandated({
            workroomId: params.workroomId,
            threadId: params.threadId,
            operationId: id,
            ...(params.mandateId ? { mandateId: params.mandateId } : {}),
            ...(params.rationale ? { rationale: params.rationale } : {}),
            participantPeerIds: participantIds(detail),
            payload: {
              version: 1,
              kind: "deliverable",
              artifactId: params.artifactId,
              artifactVersion: params.artifactVersion,
              artifactVersionId: params.artifactVersionId,
              ...(params.deliverableId ? { deliverableId: params.deliverableId } : {}),
              acceptanceCriteria: params.acceptanceCriteria,
              ...(params.note ? { note: params.note } : {}),
              mentions: mentions(detail, params.mentionHandles ?? []),
            },
          });
          return jsonResult({ operationId: id, result });
        },
      },
    ];

    if (context.workspaceDir) {
      tools.push({
        name: "atalk_task_submit_file",
        label: "Attach a file to an aTalk Task",
        description: "Encrypt and attach one file from the active OpenClaw workspace to a Task. Returns artifact identifiers that can be submitted as a deliverable.",
        parameters: Type.Object({
          ...common,
          filePath: Type.String({ minLength: 1 }),
          name: Type.Optional(Type.String({ minLength: 1, maxLength: 500 })),
          mimeType: Type.Optional(Type.String({ minLength: 1, maxLength: 200 })),
          title: Type.Optional(Type.String({ minLength: 1, maxLength: 500 })),
          description: Type.Optional(Type.String({ maxLength: 4_000 })),
          artifactType: Type.Optional(Type.String({ minLength: 1, maxLength: 160 })),
          artifactId: Type.Optional(uuid()),
          artifactVersion: Type.Optional(Type.Integer({ minimum: 1 })),
          mentionHandles: Type.Optional(handles()),
        }, { additionalProperties: false }),
        async execute(_toolCallId: string, params: {
          workroomId: string; threadId: string; operationId?: string; mandateId?: string; rationale?: string;
          filePath: string; name?: string; mimeType?: string; title?: string; description?: string;
          artifactType?: string; artifactId?: string; artifactVersion?: number; mentionHandles?: string[];
        }) {
          const id = operationId(params.operationId);
          const detail = await agent.workrooms.get(params.workroomId, 0, 1);
          const path = await workspaceFile(params.filePath, context.workspaceDir!);
          const result = await agent.workrooms.submitFileMandated({
            workroomId: params.workroomId,
            threadId: params.threadId,
            operationId: id,
            path,
            ...(params.mandateId ? { mandateId: params.mandateId } : {}),
            ...(params.rationale ? { rationale: params.rationale } : {}),
            ...(params.name ? { name: params.name } : {}),
            ...(params.mimeType ? { mimeType: params.mimeType } : {}),
            ...(params.title ? { title: params.title } : {}),
            ...(params.description ? { description: params.description } : {}),
            ...(params.artifactType ? { artifactType: params.artifactType } : {}),
            ...(params.artifactId ? { artifactId: params.artifactId } : {}),
            ...(params.artifactVersion ? { artifactVersion: params.artifactVersion } : {}),
            mentions: mentions(detail, params.mentionHandles ?? []),
            participantPeerIds: participantIds(detail),
          });
          return jsonResult({ operationId: id, result });
        },
      });
    }
  return tools;
}

export function registerAtalkTaskTools(api: OpenClawPluginApi): void {
  api.registerTool((context) => {
    const agent = getActiveAtalkAgent(context.agentAccountId ?? "default");
    return agent ? buildAtalkTaskTools(agent, context) : null;
  }, {
    names: [
      "atalk_task_list", "atalk_task_open", "atalk_task_message", "atalk_task_activity",
      "atalk_task_plan", "atalk_task_deliverable", "atalk_task_submit_file",
    ],
  });
}

File: integrations/hermes/src/atalk_hermes/tools.py
"""Hermes-native tools for governed aTalk Tasks/Workrooms."""

from __future__ import annotations

import json
import mimetypes
import os
import uuid
from pathlib import Path
from typing import Any


MAX_ATTACHMENT_BYTES = 100 * 1024 * 1024
_active_adapter: Any | None = None


def set_active_adapter(adapter: Any) -> None:
    global _active_adapter
    _active_adapter = adapter


def clear_active_adapter(adapter: Any) -> None:
    global _active_adapter
    if _active_adapter is adapter:
        _active_adapter = None


def _adapter() -> Any:
    if _active_adapter is None:
        raise RuntimeError("The aTalk Hermes platform is not connected")
    return _active_adapter


def _json(value: Any) -> str:
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"), default=str)


def _task_view(detail: dict[str, Any]) -> dict[str, Any]:
    workroom = detail["workroom"]
    return {
        "id": workroom["id"],
        "status": workroom["status"],
        "deadline": workroom.get("deadline"),
        "descriptor": detail["descriptor"],
        "membership": {
            "role": detail["membership"]["role"],
            "joinedAt": detail["membership"]["joinedAt"],
            "leftAt": detail["membership"].get("leftAt"),
        },
        "members": [{
            "id": item["membership"]["peerId"],
            "role": item["membership"]["role"],
            "leftAt": item["membership"].get("leftAt"),
            "handle": item.get("peer", {}).get("handle") if item.get("peer") else None,
            "displayName": item.get("peer", {}).get("displayName") if item.get("peer") else None,
            "type": item.get("peer", {}).get("type") if item.get("peer") else None,
        } for item in detail.get("members", [])],
        "threads": [{"id": item["id"], "kind": item["kind"], "createdAt": item["createdAt"]}
                    for item in detail.get("threads", [])],
        "permissions": [{
            "mandateId": item["mandate"]["mandateId"],
            "revision": item["mandate"]["revision"],
            "actorPeerId": item["mandate"]["actorPeerId"],
            "validFrom": item["mandate"]["validFrom"],
            "validUntil": item["mandate"]["validUntil"],
            "revoked": bool(item.get("revocation")),
        } for item in detail.get("latestMandates", [])],
        "approvals": [{
            "requestId": item["requestId"], "status": item["status"],
            "requiredApprovals": item["requiredApprovals"],
            "eligiblePeerIds": item["eligiblePeerIds"], "expiresAt": item.get("expiresAt"),
        } for item in detail.get("approvals", [])],
    }


def _participants(detail: dict[str, Any]) -> list[str]:
    return [str(item["membership"]["peerId"]) for item in detail.get("members", [])
            if not item["membership"].get("leftAt")]


def _peers(detail: dict[str, Any], handles: list[str]) -> list[dict[str, Any]]:
    active = [item["peer"] for item in detail.get("members", [])
              if not item["membership"].get("leftAt") and item.get("peer")]
    peers = []
    for handle in handles:
        peer = next((candidate for candidate in active if candidate.get("handle") == handle), None)
        if peer is None:
            raise ValueError(f"Active Task participant not found: {handle}")
        peers.append(peer)
    return peers


def _mentions(detail: dict[str, Any], handles: list[str]) -> list[dict[str, str]]:
    return [{
        "peerId": str(peer["id"]), "handle": str(peer["handle"]),
        "peerType": "AGENT" if peer.get("type") == "AGENT" else "HUMAN", "intent": "direct",
    } for peer in _peers(detail, handles)]


def _prepare_payload(detail: dict[str, Any], params: dict[str, Any]) -> dict[str, Any]:
    kind = str(params["kind"])
    if kind == "message":
        return {
            "version": 1, "kind": kind, "threadId": params["threadId"], "body": params["body"],
            "mentions": _mentions(detail, params.get("mentionHandles", [])),
            **({"replyToEventId": params["replyToEventId"]} if params.get("replyToEventId") else {}),
        }
    if kind == "activity":
        return {
            "version": 1, "kind": kind, "threadId": params["threadId"],
            "activityType": params["activityType"], "summary": params["summary"],
            "mentions": _mentions(detail, params.get("mentionHandles", [])),
            "sourceEventIds": params.get("sourceEventIds", []), "attributes": params.get("attributes", {}),
        }
    if kind == "plan":
        return {
            "version": 1, "kind": kind, "planVersion": params["planVersion"], "summary": params["summary"],
            **({"planId": params["planId"]} if params.get("planId") else {}),
            "steps": [{
                "id": step["id"], "title": step["title"], "status": step["status"],
                "assignedPeerIds": [str(peer["id"]) for peer in _peers(detail, step.get("assignedHandles", []))],
                "dependsOnStepIds": step.get("dependsOnStepIds", []),
                **({"deadline": step["deadline"]} if step.get("deadline") else {}),
            } for step in params["steps"]],
        }
    if kind == "deliverable":
        return {
            "version": 1, "kind": kind, "artifactId": params["artifactId"],
            "artifactVersion": params["artifactVersion"],
            **({"artifactVersionId": params["artifactVersionId"]} if params.get("artifactVersionId") else {}),
            **({"deliverableId": params["deliverableId"]} if params.get("deliverableId") else {}),
            "acceptanceCriteria": params["acceptanceCriteria"],
            **({"note": params["note"]} if params.get("note") else {}),
            "mentions": _mentions(detail, params.get("mentionHandles", [])),
        }
    raise ValueError(f"Unsupported aTalk Task publication: {kind}")


async def _list(params: dict[str, Any], **_kwargs: Any) -> str:
    page = await _adapter()._agent.workrooms.list(params.get("cursor"), int(params.get("limit", 50)))
    return _json({"tasks": [_task_view(item) for item in page.get("workrooms", [])],
                  "nextCursor": page.get("nextCursor")})


async def _open(params: dict[str, Any], **_kwargs: Any) -> str:
    detail = await _adapter()._agent.workrooms.get(str(params["workroomId"]), 0, 1)
    return _json(_task_view(detail))


async def _publish(params: dict[str, Any], **_kwargs: Any) -> str:
    adapter = _adapter()
    detail = await adapter._agent.workrooms.get(str(params["workroomId"]), 0, 1)
    operation_id = str(params.get("operationId") or uuid.uuid4())
    result = await adapter._agent.workrooms.publish_mandated({
        "workroomId": params["workroomId"], "threadId": params["threadId"],
        "operationId": operation_id, "payload": _prepare_payload(detail, params),
        "participantPeerIds": _participants(detail),
        **({"mandateId": params["mandateId"]} if params.get("mandateId") else {}),
        **({"rationale": params["rationale"]} if params.get("rationale") else {}),
    })
    return _json({"operationId": operation_id, "result": result})


def _workspace_file(value: str, kwargs: dict[str, Any]) -> Path:
    hinted = next((kwargs.get(key) for key in ("cwd", "workspace_dir", "working_dir")
                   if isinstance(kwargs.get(key), str) and kwargs.get(key)), None)
    root = Path(hinted or os.getenv("TERMINAL_CWD") or Path.cwd()).expanduser().resolve()
    candidate = Path(value).expanduser()
    path = (candidate if candidate.is_absolute() else root / candidate).resolve(strict=True)
    try:
        path.relative_to(root)
    except ValueError as error:
        raise ValueError("Task files must stay inside the active Hermes workspace") from error
    if not path.is_file():
        raise ValueError("Task attachment path is not a regular file")
    if path.stat().st_size > MAX_ATTACHMENT_BYTES:
        raise ValueError("aTalk Task attachments cannot exceed 100 MB")
    return path


async def _submit_file(params: dict[str, Any], **kwargs: Any) -> str:
    adapter = _adapter()
    detail = await adapter._agent.workrooms.get(str(params["workroomId"]), 0, 1)
    operation_id = str(params.get("operationId") or uuid.uuid4())
    path = _workspace_file(str(params["filePath"]), kwargs)
    result = await adapter._agent.workrooms.submit_file_mandated({
        "workroomId": params["workroomId"], "threadId": params["threadId"],
        "operationId": operation_id, "filePath": path,
        "name": params.get("name") or path.name,
        "mimeType": params.get("mimeType") or mimetypes.guess_type(path.name)[0] or "application/octet-stream",
        "mentions": _mentions(detail, params.get("mentionHandles", [])),
        "participantPeerIds": _participants(detail),
        **{key: params[key] for key in ("mandateId", "rationale", "title", "description", "artifactType",
                                        "artifactId", "artifactVersion") if params.get(key)},
    })
    return _json({"operationId": operation_id, "result": result})


def _schema(name: str, description: str, properties: dict[str, Any], required: list[str]) -> dict[str, Any]:
    return {
        "name": name, "description": description,
        "parameters": {"type": "object", "properties": properties, "required": required,
                       "additionalProperties": False},
    }


_UUID = {"type": "string", "format": "uuid"}
_HANDLE_LIST = {"type": "array", "items": {"type": "string", "pattern": "^@[a-z0-9][a-z0-9._-]{1,62}$"},
                "maxItems": 100}
_COMMON = {
    "workroomId": _UUID, "threadId": _UUID, "operationId": _UUID, "mandateId": _UUID,
    "rationale": {"type": "string", "maxLength": 4000},
}


def register_tools(ctx: Any) -> None:
    definitions = [
        ("atalk_task_list", _schema("atalk_task_list", "List encrypted aTalk Tasks assigned to this agent.",
                                    {"cursor": {"type": "string"}, "limit": {"type": "integer", "minimum": 1,
                                                                                 "maximum": 200}}, []), _list),
        ("atalk_task_open", _schema("atalk_task_open", "Open a verified, locally decrypted aTalk Task.",
                                    {"workroomId": _UUID}, ["workroomId"]), _open),
        ("atalk_task_message", _schema("atalk_task_message", "Publish an encrypted Task message and optionally direct it by @handle.", {
            **_COMMON, "kind": {"const": "message"}, "body": {"type": "string", "maxLength": 200000},
            "mentionHandles": _HANDLE_LIST, "replyToEventId": _UUID,
        }, ["workroomId", "threadId", "kind", "body"]), _publish),
        ("atalk_task_activity", _schema("atalk_task_activity", "Publish encrypted progress under the agent's current permission.", {
            **_COMMON, "kind": {"const": "activity"}, "activityType": {"type": "string", "maxLength": 160},
            "summary": {"type": "string", "maxLength": 4000}, "mentionHandles": _HANDLE_LIST,
        }, ["workroomId", "threadId", "kind", "activityType", "summary"]), _publish),
        ("atalk_task_plan", _schema("atalk_task_plan", "Publish a versioned Task plan. This requires plan.update permission.", {
            **_COMMON, "kind": {"const": "plan"}, "planId": _UUID,
            "planVersion": {"type": "integer", "minimum": 1}, "summary": {"type": "string", "maxLength": 2000},
            "steps": {"type": "array", "minItems": 1, "maxItems": 500, "items": {"type": "object", "properties": {
                "id": {"type": "string"}, "title": {"type": "string", "maxLength": 500},
                "status": {"type": "string", "enum": ["executing", "waiting_approval", "blocked", "completed", "cancelled", "expired"]},
                "assignedHandles": _HANDLE_LIST, "dependsOnStepIds": {"type": "array", "items": {"type": "string"}, "maxItems": 100},
                "deadline": {"type": "string", "format": "date-time"},
            }, "required": ["id", "title", "status"], "additionalProperties": False}},
        }, ["workroomId", "threadId", "kind", "planVersion", "summary", "steps"]), _publish),
        ("atalk_task_deliverable", _schema("atalk_task_deliverable", "Submit an encrypted artifact version for human review.", {
            **_COMMON, "kind": {"const": "deliverable"}, "artifactId": _UUID,
            "artifactVersion": {"type": "integer", "minimum": 1}, "artifactVersionId": _UUID,
            "deliverableId": _UUID, "acceptanceCriteria": {"type": "array", "minItems": 1, "maxItems": 100,
                                                              "items": {"type": "string", "maxLength": 1000}},
            "note": {"type": "string", "maxLength": 4000}, "mentionHandles": _HANDLE_LIST,
        }, ["workroomId", "threadId", "kind", "artifactId", "artifactVersion", "artifactVersionId", "acceptanceCriteria"]), _publish),
        ("atalk_task_submit_file", _schema("atalk_task_submit_file", "Encrypt and attach a file from the active Hermes workspace to a Task.", {
            **_COMMON, "filePath": {"type": "string"}, "name": {"type": "string", "maxLength": 500},
            "mimeType": {"type": "string", "maxLength": 200}, "title": {"type": "string", "maxLength": 500},
            "description": {"type": "string", "maxLength": 4000}, "artifactType": {"type": "string", "maxLength": 160},
            "artifactId": _UUID, "artifactVersion": {"type": "integer", "minimum": 1}, "mentionHandles": _HANDLE_LIST,
        }, ["workroomId", "threadId", "filePath"]), _submit_file),
    ]
    for name, schema, handler in definitions:
        ctx.register_tool(name=name, toolset="atalk", schema=schema, handler=handler, is_async=True,
                          description=schema["description"], emoji="🔐")
